File: kirun-js/src/engine/function/system/math/MathFunctionRepository.ts
import { Schema } from '../../../json/schema/Schema';
import { SchemaType } from '../../../json/schema/type/SchemaType';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { Repository } from '../../../Repository';
import { AbstractFunction } from '../../AbstractFunction';
import { Function } from '../../Function';
import { Add } from './Add';
import { GenericMathFunction } from './GenericMathFunction';
import { Hypotenuse } from './Hypotenuse';
import { Maximum } from './Maximum';
import { Minimum } from './Minimum';
import { Random } from './Random';
import { RandomAny } from './RandomAny';

export class MathFunctionRepository implements Repository<Function> {
    private readonly functionObjectsIndex: { [key: string]: AbstractFunction };
    private readonly filterableNames: string[];

    public constructor() {
        this.functionObjectsIndex = {
            Absolute: new GenericMathFunction(
                'Absolute',
                (v) => Math.abs(v),
                1,
                SchemaType.INTEGER,
                SchemaType.LONG,
                SchemaType.FLOAT,
                SchemaType.DOUBLE,
            ),
            ArcCosine: new GenericMathFunction('ArcCosine', (v) => Math.acos(v)),
            ArcSine: new GenericMathFunction('ArcSine', (v) => Math.asin(v)),
            ArcTangent: new GenericMathFunction('ArcTangent', (v) => Math.atan(v)),
            Ceiling: new GenericMathFunction('Ceiling', (v) => Math.ceil(v)),
            Cosine: new GenericMathFunction('Cosine', (v) => Math.cos(v)),
            HyperbolicCosine: new GenericMathFunction('HyperbolicCosine', (v) => Math.cosh(v)),
            CubeRoot: new GenericMathFunction('CubeRoot', (v) => Math.cbrt(v)),
            Exponential: new GenericMathFunction('Exponential', (v) => Math.exp(v)),
            ExponentialMinus1: new GenericMathFunction('ExponentialMinus1', (v) => Math.expm1(v)),
            Floor: new GenericMathFunction('Floor', (v) => Math.floor(v)),
            LogNatural: new GenericMathFunction('LogNatural', (v) => Math.log(v)),
            Log10: new GenericMathFunction('Log10', (v) => Math.log10(v)),
            Round: new GenericMathFunction(
                'Round',
                (v) => Math.round(v),
                1,
                SchemaType.INTEGER,
                SchemaType.LONG,
            ),
            Sine: new GenericMathFunction('Sine', (v) => Math.sin(v)),
            HyperbolicSine: new GenericMathFunction('HyperbolicSine', (v) => Math.sinh(v)),
            Tangent: new GenericMathFunction('Tangent', (v) => Math.tan(v)),
            HyperbolicTangent: new GenericMathFunction('HyperbolicTangent', (v) => Math.tanh(v)),
            ToDegrees: new GenericMathFunction('ToDegrees', (v) => v * (Math.PI / 180)),
            ToRadians: new GenericMathFunction('ToRadians', (v) => v * (180 / Math.PI)),
            SquareRoot: new GenericMathFunction('SquareRoot', (v) => Math.sqrt(v)),
            ArcTangent2: new GenericMathFunction('ArcTangent2', (v1, v2) => Math.atan2(v1, v2!), 2),
            Power: new GenericMathFunction('Power', (v1, v2) => Math.pow(v1, v2!), 2),
            Add: new Add(),
            Hypotenuse: new Hypotenuse(),
            Maximum: new Maximum(),
            Minimum: new Minimum(),
            Random: new Random(),
            RandomFloat: new RandomAny(
                'RandomFloat',
                Parameter.of(
                    RandomAny.MIN_VALUE,
                    Schema.ofFloat(RandomAny.MIN_VALUE).setDefaultValue(0),
                ),
                Parameter.of(
                    RandomAny.MAX_VALUE,
                    Schema.ofFloat(RandomAny.MAX_VALUE).setDefaultValue(2147483647),
                ),
                Schema.ofFloat(RandomAny.VALUE),
                (min, max) => Math.random() * (max - min) + min,
            ),
            RandomInt: new RandomAny(
                'RandomInt',
                Parameter.of(
                    RandomAny.MIN_VALUE,
                    Schema.ofInteger(RandomAny.MIN_VALUE).setDefaultValue(0),
                ),
                Parameter.of(
                    RandomAny.MAX_VALUE,
                    Schema.ofInteger(RandomAny.MAX_VALUE).setDefaultValue(2147483647),
                ),
                Schema.ofInteger(RandomAny.VALUE),
                (min, max) => Math.round(Math.random() * (max - min) + min),
            ),
            RandomLong: new RandomAny(
                'RandomLong',
                Parameter.of(
                    RandomAny.MIN_VALUE,
                    Schema.ofLong(RandomAny.MIN_VALUE).setDefaultValue(0),
                ),
                Parameter.of(
                    RandomAny.MAX_VALUE,
                    Schema.ofLong(RandomAny.MAX_VALUE).setDefaultValue(Number.MAX_SAFE_INTEGER),
                ),
                Schema.ofLong(RandomAny.VALUE),
                (min, max) => Math.round(Math.random() * (max - min) + min),
            ),
            RandomDouble: new RandomAny(
                'RandomDouble',
                Parameter.of(
                    RandomAny.MIN_VALUE,
                    Schema.ofDouble(RandomAny.MIN_VALUE).setDefaultValue(0),
                ),
                Parameter.of(
                    RandomAny.MAX_VALUE,
                    Schema.ofDouble(RandomAny.MAX_VALUE).setDefaultValue(Number.MAX_VALUE),
                ),
                Schema.ofDouble(RandomAny.VALUE),
                (min, max) => Math.random() * (max - min) + min,
            ),
        };

        this.filterableNames = Object.values(this.functionObjectsIndex).map((e) =>
            e.getSignature().getFullName(),
        );
    }

    public async find(namespace: string, name: string): Promise<Function | undefined> {
        if (namespace != Namespaces.MATH) return Promise.resolve(undefined);

        return Promise.resolve(this.functionObjectsIndex[name]);
    }

    public async filter(name: string): Promise<string[]> {
        return Promise.resolve(
            this.filterableNames.filter((e) => e.toLowerCase().indexOf(name.toLowerCase()) !== -1),
        );
    }
}

File: kirun-py/src/kirun_py/function/system/math/math_function_repository.py
from __future__ import annotations

import math
import random
import sys
from typing import Dict, List, Optional

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.function.system.math.add import Add
from kirun_py.function.system.math.generic_math_function import GenericMathFunction
from kirun_py.function.system.math.hypotenuse import Hypotenuse
from kirun_py.function.system.math.maximum import Maximum
from kirun_py.function.system.math.minimum import Minimum
from kirun_py.function.system.math.random_ import Random
from kirun_py.function.system.math.random_any import RandomAny
from kirun_py.json.schema.schema import Schema
from kirun_py.json.schema.type.schema_type import SchemaType
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces


class MathFunctionRepository:

    def __init__(self) -> None:
        self._function_objects_index: Dict[str, AbstractFunction] = {
            'Absolute': GenericMathFunction(
                'Absolute',
                lambda v: abs(v),
                1,
                [SchemaType.INTEGER, SchemaType.LONG, SchemaType.FLOAT, SchemaType.DOUBLE],
            ),
            'ArcCosine': GenericMathFunction('ArcCosine', math.acos),
            'ArcSine': GenericMathFunction('ArcSine', math.asin),
            'ArcTangent': GenericMathFunction('ArcTangent', math.atan),
            'Ceiling': GenericMathFunction('Ceiling', math.ceil),
            'Cosine': GenericMathFunction('Cosine', math.cos),
            'HyperbolicCosine': GenericMathFunction('HyperbolicCosine', math.cosh),
            'CubeRoot': GenericMathFunction('CubeRoot', lambda v: math.copysign(abs(v) ** (1.0 / 3.0), v)),
            'Exponential': GenericMathFunction('Exponential', math.exp),
            'ExponentialMinus1': GenericMathFunction('ExponentialMinus1', math.expm1),
            'Floor': GenericMathFunction('Floor', math.floor),
            'LogNatural': GenericMathFunction('LogNatural', math.log),
            'Log10': GenericMathFunction('Log10', math.log10),
            'Round': GenericMathFunction(
                'Round',
                lambda v: round(v),
                1,
                [SchemaType.INTEGER, SchemaType.LONG],
            ),
            'Sine': GenericMathFunction('Sine', math.sin),
            'HyperbolicSine': GenericMathFunction('HyperbolicSine', math.sinh),
            'Tangent': GenericMathFunction('Tangent', math.tan),
            'HyperbolicTangent': GenericMathFunction('HyperbolicTangent', math.tanh),
            'ToDegrees': GenericMathFunction('ToDegrees', math.degrees),
            'ToRadians': GenericMathFunction('ToRadians', math.radians),
            'SquareRoot': GenericMathFunction('SquareRoot', math.sqrt),
            'ArcTangent2': GenericMathFunction('ArcTangent2', math.atan2, 2),
            'Power': GenericMathFunction('Power', math.pow, 2),
            'Add': Add(),
            'Hypotenuse': Hypotenuse(),
            'Maximum': Maximum(),
            'Minimum': Minimum(),
            'Random': Random(),
            'RandomFloat': RandomAny(
                'RandomFloat',
                Parameter.of(
                    'minValue',
                    Schema.of_float('minValue').set_default_value(0),
                ),
                Parameter.of(
                    'maxValue',
                    Schema.of_float('maxValue').set_default_value(2147483647),
                ),
                Schema.of_float('value'),
                lambda mn, mx: random.random() * (mx - mn) + mn,
            ),
            'RandomInt': RandomAny(
                'RandomInt',
                Parameter.of(
                    'minValue',
                    Schema.of_integer('minValue').set_default_value(0),
                ),
                Parameter.of(
                    'maxValue',
                    Schema.of_integer('maxValue').set_default_value(2147483647),
                ),
                Schema.of_integer('value'),
                lambda mn, mx: round(random.random() * (mx - mn) + mn),
            ),
            'RandomLong': RandomAny(
                'RandomLong',
                Parameter.of(
                    'minValue',
                    Schema.of_long('minValue').set_default_value(0),
                ),
                Parameter.of(
                    'maxValue',
                    Schema.of_long('maxValue').set_default_value(sys.maxsize),
                ),
                Schema.of_long('value'),
                lambda mn, mx: round(random.random() * (mx - mn) + mn),
            ),
            'RandomDouble': RandomAny(
                'RandomDouble',
                Parameter.of(
                    'minValue',
                    Schema.of_double('minValue').set_default_value(0),
                ),
                Parameter.of(
                    'maxValue',
                    Schema.of_double('maxValue').set_default_value(sys.float_info.max),
                ),
                Schema.of_double('value'),
                lambda mn, mx: random.random() * (mx - mn) + mn,
            ),
        }

        self._filterable_names: List[str] = [
            fn.get_signature().get_full_name()
            for fn in self._function_objects_index.values()
        ]

    async def find(self, namespace: str, name: str) -> Optional[AbstractFunction]:
        if namespace != Namespaces.MATH:
            return None
        return self._function_objects_index.get(name)

    async def filter(self, name: str) -> List[str]:
        lower = name.lower()
        return [n for n in self._filterable_names if lower in n.lower()]
